### PR TITLE
[data] Add GroupedData.random_sample() for group-wise sampling

### LINKED ISSUE #53296: [Data] Stratified sampling with groupby.sample
### Description

To implement `groupby.sample()` to allow stratified sampling.

Since Ray Data is targeting large datasets and preprocessing, some of these functions are essential to provide a streamline experience without external tooling.

Pandas:
https://pandas.pydata.org/docs/reference/api/pandas.core.groupby.DataFrameGroupBy.sample.html#pandas.core.groupby.DataFrameGroupBy.sample

Some discussions on polars:
https://github.com/pola-rs/polars/issues/2661

### Use case

* simple downsampling of the full dataset for testing
* working with imbalanced dataset where we want to up or down sample different groups

### LINKED ISSUE #53297: [Data] Stratification in train_test_split
### Description

To implement `stratify` option in `Dataset.train_test_split`

Currently, the non-stratified split is implemented random shuffle + split by index. The stratified version may require a groupby + train_test_split + merge back into train/test sets. It will be very great if these can be lazy.

scikit-learn:
* https://scikit-learn.org/stable/modules/generated/sklearn.model_selection.train_test_split.html
* https://scikit-learn.org/stable/modules/cross_validation.html#stratification

### Use case

The use cases are similar to `groupby.sample` in https://github.com/ray-project/ray/issues/53296

There are cases we cannot train-test-split uniformly if the group distribution matters (e.g., imbalanced dataset)

### PR DESCRIPTION
**Pull request by saloni-garg**
## Why are these changes needed?

This PR introduces a new `GroupedData.random_sample(frac: float, *, random_state: Optional[int])` API that allows group-wise sampling of a random fraction of rows. It follows the same semantics and implementation pattern as the existing `Dataset.random_sample()` API.

Key properties:
- Samples approximately `frac * group_size` rows per group
- Supports reproducible sampling using `random_state`
- Compatible with both `pandas` and `pyarrow` batch formats
- Designed for use cases like stratified train-test splitting, downsampling, and A/B testing

The internal sampling logic is adapted directly from `Dataset.random_sample()` using seeded NumPy RNGs scoped by task index, ensuring distributed consistency.

## Related issue number

Closes #53296

## Checks

* [x] I've signed off every commit (by using the `-s` flag, i.e., `git commit -s`) in this PR.
* [x] I've run `scripts/format.sh` to lint the changes in this PR.
* [x] I've included any doc changes needed for [https://docs.ray.io/en/master/](https://docs.ray.io/en/master/).
* [x] I've made sure the tests are passing. Note that there might be a few flaky tests, see the recent failures at [https://flakey-tests.ray.io/](https://flakey-tests.ray.io/)

### Testing Strategy

* [ ] Unit tests: Added tests verifying:
  * Sampling a fraction of rows per group (including with `random_state`)
  * Reproducibility across multiple runs with the same seed
  * Empty datasets and single-group datasets
  * Sampling with `frac=0.0`, `1.0`, and invalid `frac` values

### CODE REVIEW

**Comment by saloni-garg** [comment]
> @saloni-garg Can you separate groupby.sample() (actually we should use `random_sample()` for consistency) and `train_test_split()` into two PRs? also, can you add some descriptions about your approach into the original issues?
> 
> `groupby.random_sample()` should be straight-forward: we can reuse `Dataset.random_sample`.

Thanks for the quick review, Kit. I have edited this PR to only include the group-wise sampling changes.

File: python/ray/data/grouped_data.py [hunk 1]
```diff
@@ -539,6 +539,72 @@ def std(
...
+        *,
+        random_state: Optional[int] = None,
+    ) -> "Dataset":
+        """Returns a new :class:`Dataset` containing a random fraction of the rows for each group.
+
+        .. note::
+
+            This method returns roughly ``frac * total_rows`` rows per group. An exact number
+            of rows isn't guaranteed.
+
```
Comment (wingkitlee0) [inline python/ray/data/grouped_data.py:555]:
```suggestion
.. note::
    This method does not shuffle elements within each group nor between groups.
```
Adding one more note to the doc..

File: python/ray/data/tests/test_all_to_all.py [hunk 2]
```diff
@@ -2217,6 +2217,111 @@ def get_node_id():
...
+    num_parts,
+    configure_shuffle_method,
+    disable_fallback_to_object_extension,
+):
+    # Test edge cases for sampling
+    def _to_batch_format(ds):
+        return ds.map_batches(lambda x: x, batch_size=None, batch_format=ds_format)
+
+    # Test empty dataset
+    empty_ds = ray.data.from_items([]).repartition(num_parts)
```
Comment (wingkitlee0) [inline python/ray/data/tests/test_all_to_all.py:2282]: empty dataset may not work with repartition..

**Review by wingkitlee0 — COMMENTED** [review]
Left some more comments

**Comment by wingkitlee0** [comment]
@richardliaw Maybe you or someone can take a look? thanks

File: python/ray/data/grouped_data.py [hunk 3]
```diff
@@ -539,6 +539,72 @@ def std(
         """
         return self._aggregate_on(Std, on, ignore_nulls=ignore_nulls, ddof=ddof)
 
+    @PublicAPI(api_group=CDS_API_GROUP)
+    def random_sample(
+        self,
+        frac: float,
+        *,
+        random_state: Optional[int] = None,
```
[the remark this answers is not in the document]
Comment (alexeykudinkin) [reply python/ray/data/grouped_data.py:547]: +1

File: python/ray/data/grouped_data.py [hunk 4]
```diff
@@ -539,6 +539,72 @@ def std(
...
+
+        Examples:
+            >>> import ray
+            >>> ds = ray.data.from_items([{"A": 1, "B": 1}, {"A": 1, "B": 2}, {"A": 2, "B": 3}])
+            >>> ds.groupby("A").random_sample(0.5).count()  # doctest: +SKIP
+            1
+            >>> ds.groupby("A").random_sample(0.5, random_state=42).take(2)  # doctest: +SKIP
+            [{'A': 1, 'B': 1}, {'A': 2, 'B': 3}]
+            >>> ds.groupby("A").random_sample(0.5, random_state=42).take(2)  # doctest: +SKIP
+            [{'A': 1, 'B': 1}, {'A': 2, 'B': 3}]
```
Comment (alexeykudinkin) [inline python/ray/data/grouped_data.py:564]: I think we'd rather not skip the tests here

File: python/ray/data/grouped_data.py [hunk 5]
```diff
@@ -539,6 +539,72 @@ def std(
...
+            ctx = TaskContext.get_current()
+
+            if "rng" in ctx.kwargs:
+                rng = ctx.kwargs["rng"]
+            elif seed is None:
+                rng = np.random.default_rng()
+                ctx.kwargs["rng"] = rng
+            else:
+                rng = np.random.default_rng([ctx.task_idx, seed])
+                ctx.kwargs["rng"] = rng
```
Comment (alexeykudinkin) [inline python/ray/data/grouped_data.py:592]: We have to set fixed seed even if the user-provided one is null (we can use task-id for it) to make sure that the tasks is producing deterministic results upon retries (if any)

Also, let's merge this 2 branches.

File: python/ray/data/grouped_data.py [hunk 6]
```diff
@@ -539,6 +539,72 @@ def std(
...
+
+            mask_idx = np.where(rng.random(len(batch)) < frac)[0]
+            if isinstance(batch, pa.Table):
+                return batch.take(mask_idx)
+            elif isinstance(batch, pd.DataFrame):
+                return batch.iloc[mask_idx, :]
```
Comment (alexeykudinkin) [inline python/ray/data/grouped_data.py:598]: Let's use BlockAccessor for that
